scalding-core/src/main/scala/com/twitter/scalding/serialization/Reader.scala

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2015 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.twitter.scalding.serialization
+
+import java.io.InputStream
+import scala.reflect.ClassTag
+import scala.collection.generic.CanBuildFrom
+
+/**
+ * This is a specialized typeclass to make it easier to implement Serializations.
+ * The specialization *should* mean that there is no boxing and if the JIT
+ * does its work, Reader should compose well (via collections, Tuple2, Option, Either)
+ */
+trait Reader[@specialized(Boolean, Byte, Short, Int, Long, Float, Double) +T] {
+  def read(is: InputStream): T
+}
+
+object Reader {
+  import JavaStreamEnrichments._
+
+  def read[@specialized(Boolean, Byte, Short, Int, Long, Float, Double) T](
+    is: InputStream)(implicit r: Reader[T]): T = r.read(is)
+  /*
+   * Instances below
+   */
+  implicit val unit: Reader[Unit] = new Reader[Unit] {
+    def read(is: InputStream) = ()
+  }
+  implicit val boolean: Reader[Boolean] = new Reader[Boolean] {
+    def read(is: InputStream) = is.readBoolean
+  }
+  implicit val byte: Reader[Byte] = new Reader[Byte] {
+    def read(is: InputStream) = is.readByte
+  }
+  implicit val short: Reader[Short] = new Reader[Short] {
+    def read(is: InputStream) = is.readShort
+  }
+  implicit val int: Reader[Int] = new Reader[Int] {
+    def read(is: InputStream) = is.readInt
+  }
+  implicit val long: Reader[Long] = new Reader[Long] {
+    def read(is: InputStream) = is.readLong
+  }
+  implicit val float: Reader[Float] = new Reader[Float] {
+    def read(is: InputStream) = is.readFloat
+  }
+  implicit val double: Reader[Double] = new Reader[Double] {
+    def read(is: InputStream) = is.readDouble
+  }
+  implicit val string: Reader[String] = new Reader[String] {
+    def read(is: InputStream) = {
+      val size = is.readSize
+      val bytes = new Array[Byte](size)
+      is.readFully(bytes)
+      new String(bytes, "UTF-8")
```

Comment:
fun fact did you know that at once point (not sure if still true)
`new String(bytes, Charsets.UTF_8)` was 40% slower than `new String(bytes, "UTF-8")`